File: src/JavaScript/Valid-Parentheses.js

/**
 * 检测括号对匹配是否合法
 * @param {String} parens 包含小括号和其他各种字符的字符串
 * @returns {Boolean}
 */
function validParentheses(parens) {
  // 用整数 stack 类比 Stack 
  let stack = 0;

  for (c of parens.split("")) {
    if (c === '(') stack++;
    if (c === ')') stack--;
    if (stack < 0) return false;
  }

  return stack === 0;
}


console.log(validParentheses("()(())"));

File: src/R/Valid-Parentheses.R
library(tidyverse)

#' @title 检查括号对的匹配
valid_parentheses <- function(parens) {
  num <- 0

  char_vector <- parens %>%
    str_split("") %>%
    unlist()

  for (char in char_vector) {
    if (char == "(") num <- num + 1
    if (char == ")") num <- num - 1
    num < 0 && return(FALSE)
  }

  num == 0
}


valid_parentheses("()(()())")
